### PR TITLE
268 search select loading states

### PR DESCRIPTION
**Pull request by GCHQ-Developer-741**
## Summary of the changes
Implemented loading states for select and search bar when using external filtering.

## Related issue
#268 

## Checklist
- [x] I have added relevant unit and visual regression tests.
- [x] I have manually accessibility tested any changes, if relevant.
- [x] I have ensured any changes match the Figma component library.

### CODE REVIEW

**Comment by GCHQ-Developer-741** [comment]
Since the deployment links didn't get added...

Web Components: https://mi6.github.io/ic-ui-kit/branches/268-search-select-loading-states/web-components/
React: https://mi6.github.io/ic-ui-kit/branches/268-search-select-loading-states/react/

File: audit-ci.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
     "GHSA-w5p7-h5w8-2hfq",
     "GHSA-7p7h-4mm5-852v",
     "GHSA-pfrx-2q88-qq97",
-    "GHSA-9c47-m6qq-7p4h"
+    "GHSA-9c47-m6qq-7p4h",
+    "GHSA-hc6q-2mpp-qw7j"
```
Comment (ad9242) [inline audit-ci.json:8]: think a rebase will remove need for this

**Review by ad9242 — CHANGES_REQUESTED** [review]

**Comment by ad9242** [comment]
Found a bit of a strange one:

- type a value in so that loading displays (and then loading error)

image

- tab focus to retry button (a bit strange that focus border is chopped off):
image

- now use ArrowUp or ArrowDown: the loading error gets the focussed option styling:
image

**Comment by jd239** [comment]
Can the commits be fixed please. There are updates in different packages but only a single web-component commit.

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 2]
```diff
@@ -159,9 +166,18 @@ export class SearchBar {
    */
   @Prop() options?: IcMenuOption[] = [];
 
+  /**
+   * If using external filtering, set a timeout for when loading takes too long.
+   */
+  @Prop() timeout?: number = null;
```
Comment (jd239) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:172]: Think we need to add `null` as a data type here as it's not just number.

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 3]
```diff
@@ -38,13 +38,17 @@ import searchIcon from "../../assets/search-icon.svg";
 
 let inputIds = 0;
 
+const LOADING_LABEL = "Loading...";
+const LOADING_ERROR_LABEL = "Loading Error";
```
Comment (jd239) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:42]: Should these two be a prop so they can be changed but have the default as they are above.

**Comment by jd239** [comment]
2 issues when using the loading stats

1.
- Type in two letters
- Trigger the loading error
- Tab to retry button and press enter
- Menu disappears
Should that keep the menu open and show loading again?

2.
- Type in two letters
- Trigger the loading error
- Tab to retry button and press space
- Loading error has focus with the action blue so initially couldn't tell that the retry button was still there.

**Review by jd239 — CHANGES_REQUESTED** [review]
Added a few comments. More to do with coding standards rather than implementation. Good work. Looks slick!

File: packages/web-components/src/components/ic-menu/ic-menu.tsx [hunk 4]
```diff
@@ -26,6 +32,7 @@ import {
   styleUrl: "ic-menu.css",
   shadow: true,
 })
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
[the remark this answers is not in the document]
Comment (GCHQ-Developer-741) [reply packages/web-components/src/components/ic-menu/ic-menu.tsx:35]: The actual Menu class. It never gets actually referenced so all the components have a warning which isn't really relevant. Thought to add this to get rid of the warning

**Comment by jd239** [comment]
Just had a thought. Is it possible to use the loading state with internal filtering?
- RE: Search bar. When typing the initial search phrase, the request is made to the back end to retrieve all the options (loading state appears) and then the internal filtering is used to filter the options
- RE: Select/Searchable. User clicks on the input/select to open the menu and then makes a request (loading state appears). Searchable will then use the internal filter.

**Review by MI6-451 — APPROVED** [review]

**Comment by jd239** [comment]
> Just had a thought. Is it possible to use the loading state with internal filtering?
> 
> * RE: Search bar. When typing the initial search phrase, the request is made to the back end to retrieve all the options (loading state appears) and then the internal filtering is used to filter the options
> * RE: Select/Searchable. User clicks on the input/select to open the menu and then makes a request (loading state appears). Searchable will then use the internal filter.

Has this been implemented as well? If so, is there a story to test this?

**Comment by ad9242** [comment]
Space is working now - thanks! Just a couple of things spotted:

1) when using searchable select eg searchable default story (https://mi6.github.io/ic-ui-kit/branches/268-search-select-loading-states/web-components/?path=/story/web-components-select--searchable-default), the select loses the focus border as you type eg
- click in field: focus is correct with drop down displayed:
image


- now type "ca" - it loses focus border (but displays correct filtered options)
image

this is the behaviour on develop branch (different options though)
image

2) the "x" clear button disappears when in the loading error state:
image

**Comment by ad9242** [comment]
sorry, a couple more issues:

1) type "es" & get the loading error\retry:
image
- now clear the clear selection "x"
- type in the field & you no longer get any suggestions

2) type into the field so that loading error\retry appears
- use keyboard to select "retry"
- options are updated, but the input has lost focus so you cannot select anything with arrow keys

**Review by ad9242 — APPROVED** [review]

**Review by MI6-451 — APPROVED** [review]

**Review by ad9242 — APPROVED** [review]

**Review by MI6-451 — APPROVED** [review]

**Review by jd239 — APPROVED** [review]

**Pull request merged**